### PR TITLE
Serialization -  refactor and handle short reads

### PR DESCRIPTION
**Pull request by DvirDukhan**
!!! WIP !!! DO NOT REVIEW OR MERGE UNTIL READY

This PR scope:
- [x] Decouple RedisAI logical object API from their RedisModuleType API. Introduce new pluggable encoding version management.
  - [x] Tensor
  - [x] Model
  - [x] Script
- [x] Handle short RDB reads  
  - [x] Tensor
  - [x] Model
  - [x] Script
- [ ] Flow test for short reads

### CODE REVIEW

File: src/serialization/RDB/decoder/current/v1/decode_v1.c [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+
+    RAI_Tensor *ret = RedisModule_Calloc(1, sizeof(*ret));
+    ret->tensor = (DLManagedTensor){.dl_tensor = (DLTensor){.ctx = ctx,
+                                                            .data = data,
+                                                            .ndim = ndims,
+                                                            .dtype = dtype,
+                                                            .shape = shape,
+                                                            .strides = strides,
+                                                            .byte_offset = 0},
+                                    .manager_ctx = NULL,
```
Comment (alonre24) [inline src/serialization/RDB/decoder/current/v1/decode_v1.c:59]: Why the byte_offset field is initialized to 0 and not to the loaded byte_offset?

File: src/serialization/AOF/rai_aof_rewrite.c [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    long long ndims = RAI_TensorNumDims(tensor);
+
+    RedisModuleString *dims[ndims];
+
+    for (long long i = 0; i < ndims; i++) {
+        dims[i] = RedisModule_CreateStringFromLongLong(RedisModule_GetContextFromIO(aof),
+                                                       RAI_TensorDim(tensor, i));
+    }
+
+    RedisModule_EmitAOF(aof, "AI.TENSORSET", "scvcb", key, dtypestr, dims, ndims, "BLOB", data,
```
Comment (alonre24) [inline src/serialization/AOF/rai_aof_rewrite.c:21]: What does the "scvcb" stands for?

File: src/serialization/AOF/rai_aof_rewrite.c [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+                                RedisModule_CreateString(ctx, buffer + i * chunk_size, chunk_len));
+    }
+
+    if (buffer) {
+        RedisModule_Free(buffer);
+    }
+
+    const char *backendstr = RAI_BackendName(model->backend);
+
+    RedisModule_EmitAOF(aof, "AI.MODELSET", "slccclclcvcvcv", key, backendstr, model->devicestr,
```
Comment (alonre24) [inline src/serialization/AOF/rai_aof_rewrite.c:80]: Here also, I couldn't understand what is "slccclclcvcvcv" (looks very random...)

File: src/serialization/AOF/rai_aof_rewrite.c [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+    for (size_t i = 0; i < n_chunks; i++) {
+        RedisModule_FreeString(ctx, buffers_[i]);
+    }
+    array_free(buffers_);
+}
+
+void RAI_AOFRewriteScript(RedisModuleIO *aof, RedisModuleString *key, void *value) {
+    RAI_Script *script = (RAI_Script *)value;
+    RedisModule_EmitAOF(aof, "AI.SCRIPTSET", "scccc", key, script->devicestr, script->tag, "SOURCE",
```
Comment (alonre24) [inline src/serialization/AOF/rai_aof_rewrite.c:103]: Same here, not clear what is "scccc"

File: src/serialization/AOF/rai_aof_rewrite.c [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+                        script->scriptdef);
+}
```
Comment (alonre24) [inline src/serialization/AOF/rai_aof_rewrite.c:105]: new line

**Review by alonre24 — COMMENTED** [review]
I added few minor comments and a few questions

File: src/serialization/RDB/decoder/current/v1/decode_v1.c [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+        goto cleanup;
+
+    RAI_Tensor *ret = RedisModule_Calloc(1, sizeof(*ret));
+    ret->tensor = (DLManagedTensor){.dl_tensor = (DLTensor){.ctx = ctx,
+                                                            .data = data,
+                                                            .ndim = ndims,
+                                                            .dtype = dtype,
+                                                            .shape = shape,
+                                                            .strides = strides,
+                                                            .byte_offset = 0},
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply src/serialization/RDB/decoder/current/v1/decode_v1.c:58]: from docs:
```
  /*! \brief The offset in bytes to the beginning pointer to data */
  uint64_t byte_offset;
```

File: src/serialization/AOF/rai_aof_rewrite.c (hunk 2 above)
Comment (DvirDukhan) [reply src/serialization/AOF/rai_aof_rewrite.c:21]: see `robj **moduleCreateArgvFromUserFormat(const char *cmdname, const char *fmt, int *argcp, int *flags, va_list ap)` in Redis repo

One input of the function `RedisModule_EmitAOF` is `fmt` which is the string `scvcb`, after this input you have var args input, and the  `RedisModule_EmitAOF function calls `moduleCreateArgvFromUserFormat` to know how to handle the inputs, according to the chars in `fmt`

File: src/serialization/RDB/decoder/current/v1/decode_v1.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (lantiga) [reply src/serialization/RDB/decoder/current/v1/decode_v1.c:58]: We could do this indeed. Right now every tensor that has been set from the client will have byte offset of 0, but let's load it from the serialization, it's a good idea.

**Review by lantiga — CHANGES_REQUESTED** [review]
I agree on reading / writing the byte_offset.
One note: I'm not sure the definition of the data types should be under the `serialization` directory. Let's take the data type out of that and into a dedicated directory.
Also, coverage is giving several warnings, we should make sure we are testing those functions for the most part.

**Review by lantiga — DISMISSED** [review]
LGTM. Just as a heads up (I'm sure you are aware of this), with the binary strings PR we have changed tag to be a binary string, so older RDB's won't load.

**Review by lantiga — APPROVED** [review]
LGTM. How do we test encode/decode from previous versions in CI?

**Comment by DvirDukhan** [comment]
@lantiga You are right
I will add the required tests for v0

File: src/serialization/RDB/decoder/previous/v0/decode_v0.c [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+    assert(ctx.device_type == kDLCPU);
+    assert(ctx.device_id == 0);
+
+    DLDataType dtype;
+    dtype.bits = RedisModule_LoadUnsigned(io);
+    dtype.code = RedisModule_LoadUnsigned(io);
+    dtype.lanes = RedisModule_LoadUnsigned(io);
+
+    size_t ndims = RedisModule_LoadUnsigned(io);
+    if (RedisModule_IsIOError(io))
```
Comment (alonre24) [inline src/serialization/RDB/decoder/previous/v0/decode_v0.c:24]: Why do you decide to check for IOError specifically in lines 24 and 41?

File: src/serialization/RDB/decoder/previous/v0/decode_v0.c [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+    }
+
+    RAI_ModelOpts opts = {
+        .batchsize = batchsize,
+        .minbatchsize = minbatchsize,
+        .backends_intra_op_parallelism = getBackendsIntraOpParallelism(),
+        .backends_inter_op_parallelism = getBackendsInterOpParallelism(),
+    };
+
+    len;
```
Comment (alonre24) [inline src/serialization/RDB/decoder/previous/v0/decode_v0.c:112]: What is the meaning of this declaration (line 112)?

File: src/serialization/RDB/decoder/previous/v0/decode_v0.c [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+cleanup:
+    if (devicestr)
+        RedisModule_Free(devicestr);
+    if (scriptdef)
+        RedisModule_Free(scriptdef);
+    if (tag)
+        RedisModule_FreeString(NULL, tag);
+
+    return NULL;
+}
```
Comment (alonre24) [inline src/serialization/RDB/decoder/previous/v0/decode_v0.c:245]: new line

File: tests/flow/test_serializations.py [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
...0I\x90\x8cf\x86\xe1\x1f\xbc\xf0]\x9c\xbd\x05\xcf\xc1i[IzU\x8e\x0e\x95U\xbd\xbb\xb4\xdcI]\xa7!\xac\xb9\x00\xa1\xed\x9d\xd9\x1f:\x1bx#r`9\x94\xdb\xfd\x14\x06,w7\xdb\x01\x83\x1d\x94\xa9I\x8a\xb5o\r\x15\xaaS-kh\x15\xff0s\\\x8df\x81G<\xf9\xb7\x1f\x19\x07|et\xbf\xe8\x9cY\xd2a\x08\x04\xa7\x94\x1a\x02\x97!\x04\x00\xb2 \x03A\x08\xe2\rf\x02\x18\x00#\xe1\x05\x08\x07nstants.`\xfa\x00\x1f\xe0\x12\xfa`\x00\x03\x80\x02).Au\x03m/\tW a\x00\x00@\x03\xe0\x11l\x02\x13\x00;\xe0\x03l\x03data\x80g\x007\xe0\x17g\xe0\x0f\x00\x02\x80\x02c\xe2\x00\xc2\x10\nMyModule\nq\x00)\x81}(X#U\x0f\x00trainingq\x01\x88ubq\x02`\xac\x04z\xb8\x18\x811 \x1b@\x03\x02PK\x01C5#0\x83\x82\xe3\x03J\x00\x12 \x17\xe0\x05\x00\xe3\t\x90\x80?\xe3\x01p\xe2\x03~\x00\x1c\xe0\x04<\x00R \r\xe0\x02?\xe3\x08~\xe0\x07I\xe1\x03\xbf\x00& ;\xe0\x01\x00\x01^\x01\xe0\x15I\xe2\x01\xbc\x80S\xe0\x01\xdd\xe1\x03\xa6\x00\x18 \x17\xe0\x01\x00D?\xe0\x04\x9d\xe2\x02\x07\xe0\x07E\xe1\x03?\x00\x13\xe0\x01B \x00\x00\xd4!K\xe0\x02E\xc1\xe0\x04PK\x06\x06, \x1e@\x00\x02\x1e\x03-@\x06`\x00\x00\x05`\x05\xe0\x01\x07\x00e \xd8@\x00\x01\x81\x03@\x05A\x9c\x01\x06\x07 \x06\x01\x00\xe6BV \x00@\x1e\x03PK\x05\x06 \n ;\x00\x05`/@+\x01\x00\x00\x00\t\x00MQ\xab\x8e\xfdc\x97>'
```
Comment (alonre24) [inline tests/flow/test_serializations.py:78]: Are you sure that you want the model_rdb to be hard coded instead of reading it from a file?

**Review by alonre24 — APPROVED** [review]

**Pull request merged**